### PR TITLE
Implemented i32.store and i64.store

### PR DESCRIPTION
**Pull request by ColinEberhardt**
A complete implementation of  i32.store and i64.store #59 

- The allocator no longer uses the runtime memory, it instead uses an array which represents the store (as per the spec)
- Memory now uses an `ArrayBuffer`
- 
- The `store` instruction accesses this via `frame.originatingModule.memaddrs[0]`, as defined in the spec: https://webassembly.github.io/spec/core/exec/instructions.html#exec-storen

You can see it fully working here:

https://github.com/xtuc/js-webassembly-interpreter/pull/123/files#diff-6d5a2351c718a8fe03cfcfa79ca3fd90

### CODE REVIEW

File: src/interpreter/kernel/exec.js [hunk 1]
```diff
@@ -1,3 +1,5 @@
+import { debuglog } from "util";
```
Comment (xtuc) [inline src/interpreter/kernel/exec.js]: Ah yes good point. We could also use it here https://github.com/xtuc/js-webassembly-interpreter/blob/3dde1533bec38620ef95931dfa66955b9b798beb/src/interpreter/index.js#L199-L205 to trace the execution.

File: src/interpreter/kernel/memory.js [hunk 2]
```diff
@@ -6,38 +6,41 @@ export const NULL = 0x0;
 // FIXME(sven): remove this ^
 export const ptrsize = 1;
```
Comment (xtuc) [inline src/interpreter/kernel/memory.js]: I don't think this is actually used, I wanted to fake the behavior of low-level pointer but that's maybe unnecessary?

File: src/interpreter/kernel/exec.js [hunk 3]
```diff
@@ -703,6 +705,21 @@ export function executeStackFrame(
         return pop1();
       }
 
+      /**
+       * Memory operations
+       */
+      case "store": {
+        const memAddr = frame.originatingModule.memaddrs[0];
```
Comment (xtuc) [inline src/interpreter/kernel/exec.js]: Is the memory of the module always at index 0?

File: src/interpreter/kernel/exec.js [hunk 4]
```diff
@@ -703,6 +705,21 @@ export function executeStackFrame(
...
+        const memory = (frame.allocator.get(memAddr): Memory);
```
Comment (xtuc) [inline src/interpreter/kernel/exec.js]: In Flow this is actually not a cast, you want:

```js
const memory: Memory = frame.allocator.get(memAddr);
```

File: src/interpreter/kernel/exec.js (hunk 3 above)
Comment (ColinEberhardt) [reply src/interpreter/kernel/exec.js]: yes, it is documented in the `store` instruction here: https://webassembly.github.io/spec/core/exec/instructions.html#exec-storen

and the design docs too: http://webassembly.org/docs/modules/#linear-memory-section

File: src/interpreter/kernel/exec.js (hunk 4 above)
Comment (ColinEberhardt) [reply src/interpreter/kernel/exec.js]: thanks - need to read up a bit more on Flow :-)

File: src/interpreter/kernel/memory.js (hunk 2 above)
Comment (ColinEberhardt) [reply src/interpreter/kernel/memory.js]: That's an interesting point - I guess with the real runtime the store (which the allocator uses) is a proper heap, sorting the funcs, globals etc ... whereas the allocator in this emulator stores references, rather that the data structures themselves. 

I do like the way the emulator closely mimics the real runtime.

File: src/interpreter/kernel/exec.js (hunk 3 above)
Comment (xtuc) [reply src/interpreter/kernel/exec.js]: Right, thanks for the clarification.

File: src/interpreter/runtime/values/module.js [hunk 5]
```diff
@@ -92,8 +92,26 @@ export function createInstance(
     },
 
     Memory({ node }: NodePath<Memory>) {
-      // TODO(sven): implement exporting a Memory instance
-      const memoryinstance = null;
+      const limits = {
```
Comment (ColinEberhardt) [inline src/interpreter/runtime/values/module.js]: @xtuc it looks like limits are represented as strings within the AST - is that right? should they be `number`? also they should be able to handle hex representation

Comment (xtuc) [reply src/interpreter/runtime/values/module.js]: If that's the case it's a bug, they should be number (or hex).

They are numbers in the AST
https://github.com/xtuc/js-webassembly-interpreter/blob/master/src/types/AST.js#L203-L204

Comment (xtuc) [reply src/interpreter/runtime/values/module.js]: You are right! https://github.com/xtuc/js-webassembly-interpreter/blob/64851bba4ca86b1fa9e277099d0c3ffe13de6bfc/test/compiler/parsing/fixtures/watf/table/basic/expected.json#L21-L22

It's a bug.

File: src/compiler/parsing/watf/grammar.js [hunk 6]
```diff
@@ -159,11 +160,11 @@ export function parse(tokensList: Array<Object>, source: string): Program {
         );
       }
 
-      const limits = t.limits(token.value);
+      const limits = t.limits(parse32I(token.value));
```
Comment (ColinEberhardt) [inline src/compiler/parsing/watf/grammar.js]: @xtuc - I've added a fix for the 'limits' bug. Not sure whether I should be importing numeric parsing logic into `grammar.js` - or whether `t.limits` should take string args and perform this parsing logic. Any thoughts?

Comment (xtuc) [reply src/compiler/parsing/watf/grammar.js]: Conceptually t.* shouldn't have too much logic, they're just there to construct the AST.

Comment (xtuc) [reply src/compiler/parsing/watf/grammar.js]: Conceptually t.* shouldn't have too much logic, they're just there to construct the AST.

Comment (xtuc) [reply src/compiler/parsing/watf/grammar.js]: Conceptually t.* shouldn't have too much logic, they're just there to construct the AST.

**Pull request merged**